cmd/firewall_delete.go

```diff
@@ -39,15 +61,42 @@ var firewallDeleteCmd = &cobra.Command{
 				continue
 			}
 
+			if force {
+				for _, r := range sg.IngressRule {
+					rulesTask = append(rulesTask, task{
+						&egoscale.RevokeSecurityGroupIngress{
+							ID: r.RuleID,
+						},
+						fmt.Sprintf("deleting %q rule from %q", r.RuleID, sg.Name),
```

Comment:
Deleting a SG without using `--force` doesn't print rules:

```console
$ ./exo firewall delete testsg
[+] Are you sure you want to delete the security group: "testsg" [yN]: y
delete "testsg" SG ⠹                                                                                done!

$ ./exo firewall delete --force testsg
deleting "ef79f14f-3191-40ad-8853-bd984b82d128" rule from "testsg" ⠇                                              done!
deleting "75a636ce-25fb-4839-a627-610686bf3038" rule from "testsg" ⠇                                              done!
delete "testsg" SG ⠙                                                                                done!
```

I'd suggest not printing them in `--force` mode either, since it can result in a lot of content printed in the terminal (or maybe consider an optional `--verbose` flag if you feel this is important).

Reply:
without `--force` it doesn't reove rules inside. But why not a verbose mode 👍

Reply:
Yes it does:

```console
$ exo version
exo 1.3.0 0fa5bc2 (egoscale 0.17.2)

$ exo firewall show testsg
┼─────────┼────────────────┼──────────┼─────────────┼─────────────┼──────────────────────────────────────┼
│  TYPE   │     SOURCE     │ PROTOCOL │    PORT     │ DESCRIPTION │                  ID                  │
┼─────────┼────────────────┼──────────┼─────────────┼─────────────┼──────────────────────────────────────┼
│ INGRESS │ CIDR 0.0.0.0/0 │ tcp      │ 22 (ssh)    │             │ d7683262-f56d-4897-9254-af8ee283495c │
│         │ CIDR 0.0.0.0/0 │ icmp     │ 8, 0 (Echo) │             │ 04135a21-8a67-4d86-ae1c-3c66b895054e │
┼─────────┼────────────────┼──────────┼─────────────┼─────────────┼──────────────────────────────────────┼

$ exo firewall delete testsg
[+] Are you sure you want to delete the security group: "testsg" [yN]: y
delete "testsg" SG ⠙                                                       done!
```

Reply:
💯 
```
➜  ~ exo version
exo 1.3.0 0fa5bc2 (egoscale 0.17.2)
➜  ~ exo firewall ls
┼───────────────────────────────┼────────────────────────┼──────────────────────────────────────┼
│             NAME              │      DESCRIPTION       │                  ID                  │
┼───────────────────────────────┼────────────────────────┼──────────────────────────────────────┼
│ my-exoscale-cluster-node      │                        │ d7d61189-f36b-4533-b3b1-fbc4394b8606 │
│ my-exoscale-cluster-master    │                        │ 37903ad1-eee1-44bf-bfe5-e152ef1ee5ec │
│ default                       │ Default Security Group │ b1b05d21-11de-4c38-804e-c9bdacdaaa70 │
│ my-exoscale-cluster-pv-node   │                        │ 96dc22d3-ab3f-4b9b-8398-52c41ee0b2cd │
│ my-exoscale-cluster-pv-master │                        │ d4102cfc-47de-4265-a598-9141184c8fd2 │
┼───────────────────────────────┼────────────────────────┼──────────────────────────────────────┼
➜  ~ exo firewall delete my-exoscale-cluster-node
[+] Are you sure you want to delete the security group: "my-exoscale-cluster-node" [yN]: y
delete "my-exoscale-cluster-node" SG ⠙                                                          done!
failure delete "my-exoscale-cluster-node" SG: API error ResourceInUseError 536 (ServerAPIException 9999): Cannot delete group when there are security rules that allow this group
➜  ~
```

Reply:
🤔

Reply:
😺

Reply:
```
➜  cli git:(firewall-del-all) exo firewall show my-exoscale-cluster-node
┼─────────┼───────────────────────────────┼──────────┼─────────────┼───────────────────┼──────────────────────────────────────┼
│  TYPE   │            SOURCE             │ PROTOCOL │    PORT     │    DESCRIPTION    │                  ID                  │
┼─────────┼───────────────────────────────┼──────────┼─────────────┼───────────────────┼──────────────────────────────────────┼
│ INGRESS │ SG my-exoscale-cluster-master │ ipip     │ 0           │ Calico IPIP       │ 65fba888-cbc1-4857-a2f4-d906b9f6bb70 │
│         │ SG my-exoscale-cluster-node   │ ipip     │ 0           │ Calico IPIP       │ 69c1aa7d-2da0-40e8-9ea5-a9fe386a834d │
│         │ SG my-exoscale-cluster-master │ tcp      │ 10250       │ Kubelet API       │ 3c9a02ff-f953-4fbf-9639-f3e523c57a0e │
│         │ SG my-exoscale-cluster-node   │ tcp      │ 10250       │ Kubelet API       │ ba3c453d-8613-4c27-8285-de90e37ba396 │
│         │ SG my-exoscale-cluster-master │ tcp      │ 179         │ Calico BGP        │ 1a74d68c-ef9d-417c-a269-5e2bcc2cd3f3 │
│         │ SG my-exoscale-cluster-node   │ tcp      │ 179         │ Calico BGP        │ 18625a19-1998-43a6-aae3-b6737795c5d3 │
│         │ CIDR 0.0.0.0/0                │ tcp      │ 22 (ssh)    │ SSH               │ 7e17e808-455e-48bf-a670-81b2a07c4c10 │
│         │ CIDR ::/0                     │ tcp      │ 22 (ssh)    │ SSH               │ 80a451e4-4103-4f99-b43f-9b8c37ad96da │
│         │ CIDR 0.0.0.0/0                │ tcp      │ 30000-32767 │ NodePort Services │ aa5ee3e5-e29d-44a8-9437-95bd239ba2e8 │
│         │ CIDR ::/0                     │ tcp      │ 30000-32767 │ NodePort Services │ 6d316dba-56cf-46c8-ab59-0ede2d58f94a │
┼─────────┼───────────────────────────────┼──────────┼─────────────┼───────────────────┼──────────────────────────────────────┼
➜  cli git:(firewall-del-all)
```

Reply:
```console
$ git co firewall-del-all
Switched to branch 'firewall-del-all'
Your branch is up to date with 'origin/firewall-del-all'.
$ git pull
remote: Enumerating objects: 4, done.
remote: Counting objects: 100% (4/4), done.
remote: Compressing objects: 100% (4/4), done.
remote: Total 4 (delta 0), reused 0 (delta 0), pack-reused 0
Unpacking objects: 100% (4/4), done.
From github.com:exoscale/cli
   d5e76f0..e139bab  firewall-del-all -> origin/firewall-del-all
First, rewinding head to replay your work on top of it...
Fast-forwarded firewall-del-all to e139babf360bfa4fc85a203e0f646641cfa50ccc.
$ make exo
go build -ldflags "-s -w -X main.version=dev -X main.commit=e139babf360bfa4fc85a203e0f646641cfa50ccc" -mod=vendor -o exo
$ ./exo version
exo dev e139babf360bfa4fc85a203e0f646641cfa50ccc (egoscale 0.17.2)
$ ./exo firewall show testsg
┼─────────┼────────────────┼──────────┼─────────────┼─────────────┼──────────────────────────────────────┼
│  TYPE   │     SOURCE     │ PROTOCOL │    PORT     │ DESCRIPTION │                  ID                  │
┼─────────┼────────────────┼──────────┼─────────────┼─────────────┼──────────────────────────────────────┼
│ INGRESS │ CIDR 0.0.0.0/0 │ tcp      │ 22 (ssh)    │             │ f011e383-da05-4b22-a36a-f36d4784c1dd │
│         │ CIDR 0.0.0.0/0 │ icmp     │ 8, 0 (Echo) │             │ 1e0e3e5e-31c0-4803-a4d1-daa4e1e3685d │
┼─────────┼────────────────┼──────────┼─────────────┼─────────────┼──────────────────────────────────────┼
$ ./exo firewall delete testsg
[+] Are you sure you want to delete the security group: "testsg" [yN]: y
delete "testsg" SG ⠹
```

😱

Reply:
OMG @falzm break the internet
![image](https://user-images.githubusercontent.com/15922119/57919645-7fbe0600-7899-11e9-81de-7773db3b7b07.png)

Reply:
@pierre-emmanuelJ 

You're unable to delete your security group because you have another security group referencing this security group (so you cannot delete it).

That mean that you should delete all rhe rules inside the security groups, and then security groups. That way, you will be sure that no group reference another group.

Reply:
This also means that if a user references a SG *foo* into the `default` SG you won't be able to delete the SG *foo* because it is still referenced...

Reply:
okay in this way
```
➜ exo firewall add  test minecraft
Add "minecraft" rule for "test" ⠇                                                                                done!
┼─────────┼────────────────┼──────────┼───────────────────┼─────────────┼──────────────────────────────────────┼
│  TYPE   │     SOURCE     │ PROTOCOL │       PORT        │ DESCRIPTION │                  ID                  │
┼─────────┼────────────────┼──────────┼───────────────────┼─────────────┼──────────────────────────────────────┼
│ INGRESS │ CIDR 0.0.0.0/0 │ all      │ 0                 │             │ 98e5ae47-117c-4cd8-9908-d2be9c62263b │
│         │ CIDR 0.0.0.0/0 │ tcp      │ 25565 (minecraft) │             │ 956dddef-2da9-4a99-b404-435c2a73e34d │
┼─────────┼────────────────┼──────────┼───────────────────┼─────────────┼──────────────────────────────────────┼
➜ exo firewall delete test
[+] Are you sure you want to delete the security group: "test" [yN]: y
delete "test" SG ⠙                                                                                done!
```

Reply:
Yes, but in your case if you try to delete minecraft it will fail.

Reply:
oh this is an other issue so no take part of this pull request
Can you please create an issue for that please thanks

Reply:
So @falzm @mcorbin in this PR the `--force` `--all` delete all rules from all SG and then delete all SG. So in this case  `--force` `--all` will not fail, even if an SG is referenced in an other one

Reply:
Sorry, but what I said above still stands: it's not possible to delete a SG that is referenced in the `default` SG:

```
./exo firewall create testsg1
Create security group "testsg1" ⠸ done!
┼─────────┼────┼──────────────────────────────────────┼
│  NAME   │ ID │                                      │
┼─────────┼────┼──────────────────────────────────────┼
│ testsg1 │    │ c410754b-4baa-4a41-864b-c49632586e6e │
┼─────────┼────┼──────────────────────────────────────┼

$ ./exo firewall create testsg2
Create security group "testsg2" ⠼ done!
┼─────────┼────┼──────────────────────────────────────┼
│  NAME   │ ID │                                      │
┼─────────┼────┼──────────────────────────────────────┼
│ testsg2 │    │ a10209af-d086-46e1-9b87-b9220217dfc2 │
┼─────────┼────┼──────────────────────────────────────┼

$ ./exo firewall add default --protocol tcp --port 80-80 --security-group testsg1
Add rule for "default" with port 80 ⠇done!
┼─────────┼────────────────┼──────────┼─────────────┼─────────────┼──────────────────────────────────────┼
│  TYPE   │     SOURCE     │ PROTOCOL │    PORT     │ DESCRIPTION │                  ID                  │
┼─────────┼────────────────┼──────────┼─────────────┼─────────────┼──────────────────────────────────────┼
│ INGRESS │ CIDR 0.0.0.0/0 │ tcp      │ 22 (ssh)    │             │ d4a84b20-279f-4bf3-a9c9-f66b017babb6 │
│         │ CIDR 0.0.0.0/0 │ icmp     │ 8, 0 (Echo) │             │ 70936843-fac7-443a-a92d-708ed38ac42f │
│         │ SG testsg1     │ tcp      │ 80 (http)   │             │ fbf90d33-e3f0-4af5-a6fa-b9285d502b91 │
┼─────────┼────────────────┼──────────┼─────────────┼─────────────┼──────────────────────────────────────┼

$ ./exo firewall add testsg1 --protocol tcp --port 443-443 --security-group testsg2
Add rule for "testsg1" with port 443 ⠏ done!
┼─────────┼────────────┼──────────┼─────────────┼─────────────┼──────────────────────────────────────┼
│  TYPE   │   SOURCE   │ PROTOCOL │    PORT     │ DESCRIPTION │                  ID                  │
┼─────────┼────────────┼──────────┼─────────────┼─────────────┼──────────────────────────────────────┼
│ INGRESS │ SG testsg2 │ tcp      │ 443 (https) │             │ f2de15ea-7f36-475e-a091-f3c659e398be │
┼─────────┼────────────┼──────────┼─────────────┼─────────────┼──────────────────────────────────────┼

$ ./exo firewall delete --all --force
deleting "f2de15ea-7f36-475e-a091-f3c659e398be" rule from "testsg1" ⠋ done!
delete "testsg2" SG ⠼ done!
delete "testsg1" SG ⠼ done!
failure delete "testsg1" SG: API error ResourceInUseError 536 (ServerAPIException 9999): Cannot delete group when there are security rules that allow this group
```